In [1]:
%load_ext autoreload
%autoreload 2

In [171]:
import datetime
import pandas as pd
import numpy as np
from functools import partial

from utils.datafetch import *
from utils.vectorized_funs import *
from utils.datapipe import *

> We'll need some way of loading our possible tickets.

> We'll have some information for AMEX, NASDAQ and NYSE on one csv each in the config folder

In [185]:
rr = load_exchangesinfos()

amex has 400 tickers.
nasdaq has 3081 tickers.
nyse has 3299 tickers.


Final dataset has 6780 records
(NOK) Final dataset has duplicate tickers (6779 distinct)



> We'll look into the duplicate tickers when applicable

In [30]:
_r = duplicate_tickers(rr, ign_exchange=True)
_r

Symbol                               Name     MarketCap  \
2716    MSG  The Madison Square Garden Company  5.837670e+09   
2922    MSG  The Madison Square Garden Company  5.837670e+09   

                 Sector             Industry exchange ignoring_exchange  
2716  Consumer Services  Television Services   nasdaq             False  
2922  Consumer Services  Television Services     nyse              True

> in this case the ticker is the same company appearing in two exchanges, so we'll drop one of them

In [31]:
rr = rr.drop([2922])

In [32]:
## this time this should be empty...

duplicate_tickers(rr, ign_exchange=True)

Empty DataFrame
Columns: [Symbol, Name, MarketCap, Sector, Industry, exchange, ignoring_exchange]
Index: []

> lets have a look at your datareader wrapper

In [33]:
fetch_quotes("AMZN", from_date=datetime.datetime(2016,12,1))

Open        High         Low       Close   Volume  \
Date                                                                  
2016-12-01  752.409973  753.369995  738.030029  743.650024  4626500   
2016-12-02  743.400024  748.489990  736.700012  740.340027  3499200   
2016-12-05  745.000000  761.489990  742.000000  759.359985  4298200   
2016-12-06  763.989990  768.239990  757.250000  764.719971  3788300   
2016-12-07  764.549988  770.419983  755.820007  770.419983  3655300   
2016-12-08  771.869995  773.789978  765.190002  767.330017  3185700   
2016-12-09  770.000000  770.250000  765.340027  768.659973  2458800   
2016-12-12  766.400024  766.890015  757.200012  760.119995  2956900   
2016-12-13  764.960022  782.460022  762.000000  774.340027  5244700   
2016-12-14  778.250000  780.859985  762.809998  768.820007  5406500   
2016-12-15  766.280029  769.099976  760.309998  761.000000  3794200   
2016-12-16  765.000000  765.130005  754.000000  757.770020  4828700   
2016-12-19  758.890015  770.500000  756.159973  766.000000  3103200   
2016-12-20  768.650024  774.390015  767.710022  771.219971  2667900   
2016-12-21  770.000000  771.219971  765.700012  770.599976  2040400   
2016-12-22  768.119995  771.210022  763.020020  766.340027  2518300   
2016-12-23  764.549988  766.500000  757.989990  760.590027  1976900   
2016-12-27  763.400024  774.650024  761.200012  771.400024  2627400   
2016-12-28  776.250000  780.000000  770.500000  772.130005  3279100   
2016-12-29  772.400024  773.400024  760.849976  765.150024  3153500   
2016-12-30  766.469971  767.400024  748.280029  749.869995  4125300   
2017-01-03  757.919983  758.760010  747.700012  753.669983  3518600   
2017-01-04  758.390015  759.679993  754.200012  757.179993  2495900   
2017-01-05  761.549988  782.400024  760.260010  780.450012  5813700   
2017-01-06  782.359985  799.440002  778.479980  795.989990  5958800   
2017-01-09  798.000000  801.770020  791.770020  796.919983  3440100   
2017-01-10  796.599976  798.000000  789.539978  795.900024  2556700   
2017-01-11  793.659973  799.500000  789.510010  799.020020  2982400   
2017-01-12  800.309998  814.130005  799.500000  813.640015  4861500   
2017-01-13  814.320007  821.650024  811.400024  817.140015  3773500   
2017-01-17  815.700012  816.000000  803.440002  809.719971  3659400   
2017-01-18  809.500000  811.729980  804.270020  807.479980  2336000   
2017-01-19  810.000000  813.510010  807.320007  809.039978  2529100   
2017-01-20  815.280029  816.020020  806.260010  808.330017  3338200   
2017-01-23  806.799988  818.500000  805.080017  817.880005  2680900   
2017-01-24  822.000000  823.989990  814.500000  822.440002  2965900   

             Adj Close  
Date                    
2016-12-01  743.650024  
2016-12-02  740.340027  
2016-12-05  759.359985  
2016-12-06  764.719971  
2016-12-07  770.419983  
2016-12-08  767.330017  
2016-12-09  768.659973  
2016-12-12  760.119995  
2016-12-13  774.340027  
2016-12-14  768.820007  
2016-12-15  761.000000  
2016-12-16  757.770020  
2016-12-19  766.000000  
2016-12-20  771.219971  
2016-12-21  770.599976  
2016-12-22  766.340027  
2016-12-23  760.590027  
2016-12-27  771.400024  
2016-12-28  772.130005  
2016-12-29  765.150024  
2016-12-30  749.869995  
2017-01-03  753.669983  
2017-01-04  757.179993  
2017-01-05  780.450012  
2017-01-06  795.989990  
2017-01-09  796.919983  
2017-01-10  795.900024  
2017-01-11  799.020020  
2017-01-12  813.640015  
2017-01-13  817.140015  
2017-01-17  809.719971  
2017-01-18  807.479980  
2017-01-19  809.039978  
2017-01-20  808.330017  
2017-01-23  817.880005  
2017-01-24  822.440002

> Lets see how we could prepare an initial data load

In [34]:
tkrs = dict()
errs = dict()
itr = None

for tkr in rr.iloc[0:5]["Symbol"].values.tolist():
    itr = fetch_quotes(tkr)
    if itr is None:
        errs[tkr] = itr
        print("ERROR for {}".format(tkr))
    else:
        tkrs[tkr] = itr
        print("Received {}".format(tkr))


Received XXII
Received FAX
Received IAF
Received CH
ERROR for ABE           


In [35]:
_r = initial_dataload(["AAPL", "TWTR", "XXII"])

(1/3) Recv. and Stored AAPL
(2/3) Recv. and Stored TWTR
(3/3) Recv. and Stored XXII
Took 0:00:01.425871


> lets try with the ticker dataframe now

In [10]:
ticker_list = rr.iloc[0:100]["Symbol"].values
_r = initial_dataload(ticker_list, verbose=True, del_temp=True)
ticker_list

(1/100) Recv. and Stored XXII
(2/100) Recv. and Stored FAX
(3/100) Recv. and Stored IAF
(4/100) Recv. and Stored CH
(5/100) ERROR receiving ABE           
(6/100) Recv. and Stored FCO
(7/100) Recv. and Stored IF
(8/100) Recv. and Stored ISL
(9/100) Recv. and Stored ACU
(10/100) Recv. and Stored AIII
(11/100) Recv. and Stored ATNM
(12/100) Recv. and Stored AE
(13/100) Recv. and Stored ADK
(14/100) ERROR receiving ADK^A
(15/100) Recv. and Stored ACY
(16/100) Recv. and Stored WGA
(17/100) Recv. and Stored AIRI
(18/100) ERROR receiving AA^
(19/100) ERROR receiving AXX
(20/100) Recv. and Stored AXU
(21/100) Recv. and Stored AAU
(22/100) Recv. and Stored APT
(23/100) ERROR receiving ALTV
(24/100) Recv. and Stored AAMC
(25/100) Recv. and Stored DIT
(26/100) ERROR receiving APP
(27/100) Recv. and Stored ADGE
(28/100) Recv. and Stored ALN
(29/100) Recv. and Stored AMS
(30/100) Recv. and Stored AMPE
(31/100) Recv. and Stored AXN
(32/100) Recv. and Stored AMCO
(33/100) Recv. and Stored HRT
(34/10

array(['XXII', 'FAX', 'IAF', 'CH', 'ABE           ', 'FCO', 'IF', 'ISL',
       'ACU', 'AIII', 'ATNM', 'AE', 'ADK', 'ADK^A', 'ACY', 'WGA', 'AIRI',
       'AA^', 'AXX', 'AXU', 'AAU', 'APT', 'ALTV', 'AAMC', 'DIT', 'APP',
       'ADGE', 'ALN', 'AMS', 'AMPE', 'AXN', 'AMCO', 'HRT', 'AKG', 'AINC',
       'ASB.WS', 'AST', 'AWX', 'AVL', 'ASM', 'BTG', 'BTN', 'BKJ', 'BCV',
       'BAA', 'BHB', 'BRN', 'BIR^A', 'BGSF', 'BPMX', 'BTX', 'BTX.WS',
       'BGI', 'BZM', 'MHE', 'BLE', 'BLJ', 'BFY', 'BHV', 'BDR', 'BRG',
       'BVX', 'BWL.A', 'BZC', 'BTI', 'CANF', 'ROX', 'CAW', 'CVM', 'CVM.WS',
       'CEF', 'GTU', 'CET', 'LEU', 'CCF', 'CQH', 'CQP', 'LNG', 'CVR',
       'CNR', 'CPHI', 'CKX', 'GLV', 'GLQ', 'GLO', 'CRV', 'MOC', 'CIX',
       'LODE', 'CTO', 'MCF', 'CUO', 'CMT', 'CVRS', 'CRMD', 'CRF', 'CLM',
       'CVU', 'CIK', 'DHY'], dtype=object)

^ It is nicer to have the time from start to finish of this

In [21]:
start = datetime.datetime.now()

In [22]:
end = datetime.datetime.now()

In [23]:
d = end-start

In [24]:
d.seconds

4

In [25]:
print(str(d))

0:00:04.744819


> Let's try to read back in one of the files we just created....

In [146]:
XXII = pd.read_csv("{}/{}.csv".format(DATA_PATH, "XXII"))

print("Shape: {}".format(XXII.shape))
print("\nStats\n")
print(XXII.describe())

Shape: (1509, 7)

Stats

              Open         High          Low        Close        Volume  \
count  1509.000000  1509.000000  1509.000000  1509.000000  1.509000e+03   
mean      1.235274     1.279557     1.176797     1.230465  3.093797e+05   
std       0.808300     0.840946     0.760240     0.798593  4.955789e+05   
min       0.155000     0.200000     0.150000     0.200000  0.000000e+00   
25%       0.780000     0.810000     0.740000     0.778000  1.190000e+04   
50%       1.040000     1.080000     1.000000     1.030000  1.654000e+05   
75%       1.340000     1.370000     1.280000     1.330000  4.024000e+05   
max       6.150000     6.360000     5.750000     6.200000  6.531700e+06   

         Adj Close  
count  1509.000000  
mean      1.230465  
std       0.798593  
min       0.200000  
25%       0.778000  
50%       1.030000  
75%       1.330000  
max       6.200000  


> Lets see what happens if the file does not exist....
> We might want to consider the FileNotFoundError...

In [17]:
try:
    XXIIAAAAAAAAA = pd.read_csv("{}/{}.csv".format(DATA_PATH, "XXIIAAAAAAAAA"))
except (FileNotFoundError):
    print("File not found.")
    print("We could create a function to download the data instead of erroring...")

File not found.
We could create a function to download the data instead of erroring...


> While we're here, let's look at the diference between csv and hdfs...

In [36]:
_tkrs = ["AAPL", "TWTR", "XXII", "AMZN", "GOOG"]

_r = initial_dataload(_tkrs)

for t in _tkrs:
    itr = pd.read_csv("{}/{}.csv".format(DATA_PATH, t))
    itr.to_hdf("{}/{}.h5".format(DATA_PATH, t), "stock_daily")

(1/5) Recv. and Stored AAPL
(2/5) Recv. and Stored TWTR
(3/5) Recv. and Stored XXII
(4/5) Recv. and Stored AMZN
(5/5) Recv. and Stored GOOG
Took 0:00:02.892399


In [37]:
%%bash
cd data
pwd
ls -lh
du -h

/home/jovyan/work/FintechSandbox/data
total 35M
-rw-r--r-- 1 jovyan users  59K Jan  5 19:58 AAMC.csv
-rw-r--r-- 1 jovyan users 617K Jan 25 20:02 AAPL.csv
-rw-r--r-- 1 jovyan users 3.4M Jan 25 20:02 AAPL.h5
-rw-r--r-- 1 jovyan users 114K Jan  5 19:58 AAU.csv
-rw-r--r-- 1 jovyan users 396K Jan  5 19:58 ACU.csv
-rw-r--r-- 1 jovyan users 207K Jan  5 19:58 ACY.csv
-rw-r--r-- 1 jovyan users 137K Jan  5 19:58 ADGE.csv
-rw-r--r-- 1 jovyan users 151K Jan  5 19:58 ADK.csv
-rw-r--r-- 1 jovyan users 437K Jan  5 19:58 AE.csv
-rw-r--r-- 1 jovyan users 204K Jan  5 19:58 AIII.csv
-rw-r--r-- 1 jovyan users  34K Jan  5 19:58 AINC.csv
-rw-r--r-- 1 jovyan users 172K Jan  5 19:58 AIRI.csv
-rw-r--r-- 1 jovyan users  93K Jan  5 19:58 AKG.csv
-rw-r--r-- 1 jovyan users  75K Jan  5 19:58 ALN.csv
-rw-r--r-- 1 jovyan users  70K Jan  5 19:58 AMCO.csv
-rw-r--r-- 1 jovyan users  69K Jan  5 19:58 AMPE.csv
-rw-r--r-- 1 jovyan users 365K Jan  5 19:58 AMS.csv
-rw-r--r-- 1 jovyan users 339K Jan 25 20:02 AMZN.csv
-rw-r--r

> Ok, so hdf5 will help us manage really large files. But is a bit of overkill for this kind of problem right now

In [38]:
_stock = dict()

_stock["AAPL"] = pd.read_csv("{}/{}.csv".format(DATA_PATH, "AAPL"))
_stock["TWTR"] = pd.read_csv("{}/{}.csv".format(DATA_PATH, "TWTR"))
_stock["XXII"] = pd.read_csv("{}/{}.csv".format(DATA_PATH, "XXII"))
_stock["GOOG"] = pd.read_csv("{}/{}.csv".format(DATA_PATH, "GOOG"))
_stock["AMZN"] = pd.read_csv("{}/{}.csv".format(DATA_PATH, "AMZN"))

> Now, lets settle on a little ticker list

In [39]:

_tkrs_df = pd.DataFrame(_tkrs, columns=["ticker"])
_tkrs_df

ticker
0   AAPL
1   TWTR
2   XXII
3   AMZN
4   GOOG

> Let us take a look at their min and max dates

In [40]:
_tkrs_df["min"] = _tkrs_df["ticker"].apply(lambda s: _stock[s]["Date"].min())
_tkrs_df["max"] = _tkrs_df["ticker"].apply(lambda s: _stock[s]["Date"].max())

_tkrs_df

ticker         min         max
0   AAPL  1980-12-12  2017-01-24
1   TWTR  2013-11-07  2017-01-24
2   XXII  2011-01-26  2017-01-24
3   AMZN  1997-05-15  2017-01-24
4   GOOG  2004-08-19  2017-01-24

> We have been ignoring the fact that the dates come as strings from the csv

In [41]:
_stock["GOOG"]["Date"].head()

0    2004-08-19
1    2004-08-20
2    2004-08-23
3    2004-08-24
4    2004-08-25
Name: Date, dtype: object

> Notice the **dtype** is "object" instead of "date".

> We can take care of this using **to_datetime**

In [44]:
pd.to_datetime(_stock["GOOG"]["Date"], infer_datetime_format=True).head()

0   2004-08-19
1   2004-08-20
2   2004-08-23
3   2004-08-24
4   2004-08-25
Name: Date, dtype: datetime64[ns]

In [45]:
def convert_dates(s):
    _stock[s]["Date"] = pd.to_datetime(_stock[s]["Date"], infer_datetime_format=True)

_a = _tkrs_df["ticker"].apply(convert_dates)

print(_stock["GOOG"]["Date"].head())
print("\n")
print(_stock["AAPL"]["Date"].head())

0   2004-08-19
1   2004-08-20
2   2004-08-23
3   2004-08-24
4   2004-08-25
Name: Date, dtype: datetime64[ns]


0   1980-12-12
1   1980-12-15
2   1980-12-16
3   1980-12-17
4   1980-12-18
Name: Date, dtype: datetime64[ns]


In [46]:
_stock["GOOG"].head()

Date        Open        High         Low       Close    Volume  \
0 2004-08-19  100.000168  104.060182   95.960165  100.340176  44871300   
1 2004-08-20  101.010175  109.080187  100.500174  108.310183  22942800   
2 2004-08-23  110.750191  113.480193  109.050183  109.400185  18342800   
3 2004-08-24  111.240189  111.600192  103.570177  104.870176  15319700   
4 2004-08-25  104.960181  108.000187  103.880180  106.000184   9232100   

   Adj Close  
0  50.119968  
1  54.100990  
2  54.645447  
3  52.382705  
4  52.947145

In [47]:
_tkrs_df

ticker         min         max
0   AAPL  1980-12-12  2017-01-24
1   TWTR  2013-11-07  2017-01-24
2   XXII  2011-01-26  2017-01-24
3   AMZN  1997-05-15  2017-01-24
4   GOOG  2004-08-19  2017-01-24

> From now on it is safe to slice the *Date* columns using datetime objects

In [48]:
_stock["GOOG"]["Date"].head() > datetime.datetime(2004, 8, 20)

0    False
1    False
2     True
3     True
4     True
Name: Date, dtype: bool

In [49]:
_stock["GOOG"]["Date"].head() > datetime.datetime(2004, 8, 24)

0    False
1    False
2    False
3    False
4     True
Name: Date, dtype: bool

> Lets take a look at rolling indicators

In [50]:
pd.rolling_mean(_stock["GOOG"]["Open"], 5).iloc[0:14]

/opt/conda/lib/python3.5/site-packages/ipykernel/__main__.py:1: FutureWarning: pd.rolling_mean is deprecated for Series and will be removed in a future version, replace with 
	Series.rolling(center=False,window=5).mean()
  if __name__ == '__main__':


0            NaN
1            NaN
2            NaN
3            NaN
4     105.592181
5     106.582183
6     108.000185
7     106.906183
8     105.118179
9     104.666178
10    103.514177
11    102.084174
12    101.230174
13    100.918174
Name: Open, dtype: float64

In [51]:
_stock["GOOG"]["Open"].rolling(window=5, center=False).mean().iloc[0:14]

0            NaN
1            NaN
2            NaN
3            NaN
4     105.592181
5     106.582183
6     108.000185
7     106.906183
8     105.118179
9     104.666178
10    103.514177
11    102.084174
12    101.230174
13    100.918174
Name: Open, dtype: float64

In [52]:
_stock["GOOG"]["Close"].rolling(window=5, center=False).mean().iloc[0:14]

0            NaN
1            NaN
2            NaN
3            NaN
4     105.784181
5     107.298182
6     106.866182
7     105.388179
8     104.888179
9     103.738176
10    102.458174
11    101.230172
12    101.144173
13    101.130172
Name: Close, dtype: float64

> One can slice a subset of columns using loc and then call apply across all the sliced columns.

> The result will be a new dataframe with the rolled averages row-by-row across each column

In [53]:
def apply_rolling_mean(s):
    return s.rolling(window=5, center=False).mean()

_rolling5 = _stock["GOOG"].loc[:, ["Open", "High", "Low", "Close", "Volume", "Adj Close"]].apply(apply_rolling_mean)
_rolling5.iloc[0:14]


Open        High         Low       Close      Volume  Adj Close
0          NaN         NaN         NaN         NaN         NaN        NaN
1          NaN         NaN         NaN         NaN         NaN        NaN
2          NaN         NaN         NaN         NaN         NaN        NaN
3          NaN         NaN         NaN         NaN         NaN        NaN
4   105.592181  109.244188  102.592176  105.784181  22141740.0  52.839251
5   106.582183  110.022189  104.332179  107.298182  14593200.0  53.595495
6   108.000185  109.930189  105.370180  106.866182  11252880.0  53.379711
7   106.906183  108.332187  103.962178  105.388179   8628600.0  52.641448
8   105.118179  106.754185  103.680178  104.888179   6552900.0  52.391698
9   104.666178  105.748184  102.838175  103.738176   6542800.0  51.817271
10  103.514177  104.632181  101.694174  102.458174   8155160.0  51.177909
11  102.084174  103.256179  100.420171  101.230172   7942280.0  50.564522
12  101.230174  102.558177   99.940171  101.144173   8073040.0  50.521565
13  100.918174  102.422177   99.608170  101.130172   8086640.0  50.514572

> We can concatenate the new frame and the old one using concat (we need to set the *axis=* in order to do it across columns)

> We can also reset the column names so that we don't have duplicate names (we could have done this in the previous step on the _rolling5 frame)

In [54]:
_result = pd.concat([_stock["GOOG"], _rolling5], axis=1)
_result.columns = ['Date', 'Open', 'High', 'Low', 'Close', 'Volume', 'Adj Close', 'Open_roll5','High_roll5', 'Low_roll5', 'Close_roll5', 'Volume_roll5', 'Adj Close_roll5']

_result.iloc[0:14]

Date        Open        High         Low       Close    Volume  \
0  2004-08-19  100.000168  104.060182   95.960165  100.340176  44871300   
1  2004-08-20  101.010175  109.080187  100.500174  108.310183  22942800   
2  2004-08-23  110.750191  113.480193  109.050183  109.400185  18342800   
3  2004-08-24  111.240189  111.600192  103.570177  104.870176  15319700   
4  2004-08-25  104.960181  108.000187  103.880180  106.000184   9232100   
5  2004-08-26  104.950180  107.950188  104.660179  107.910182   7128600   
6  2004-08-27  108.100185  108.620186  105.690180  106.150181   6241200   
7  2004-08-30  105.280178  105.490184  102.010172  102.010172   5221400   
8  2004-08-31  102.300173  103.710180  102.160177  102.370175   4941200   
9  2004-09-01  102.700174  102.970180   99.670169  100.250171   9181600   
10 2004-09-02   99.190173  102.370175   98.940170  101.510173  15190400   
11 2004-09-03  100.950174  101.740174   99.320168  100.010169   5176800   
12 2004-09-07  101.010175  102.000178   99.610169  101.580175   5875200   
13 2004-09-08  100.740176  103.030180  100.500174  102.300173   5009200   

    Adj Close  Open_roll5  High_roll5   Low_roll5  Close_roll5  Volume_roll5  \
0   50.119968         NaN         NaN         NaN          NaN           NaN   
1   54.100990         NaN         NaN         NaN          NaN           NaN   
2   54.645447         NaN         NaN         NaN          NaN           NaN   
3   52.382705         NaN         NaN         NaN          NaN           NaN   
4   52.947145  105.592181  109.244188  102.592176   105.784181    22141740.0   
5   53.901190  106.582183  110.022189  104.332179   107.298182    14593200.0   
6   53.022069  108.000185  109.930189  105.370180   106.866182    11252880.0   
7   50.954132  106.906183  108.332187  103.962178   105.388179     8628600.0   
8   51.133953  105.118179  106.754185  103.680178   104.888179     6552900.0   
9   50.075011  104.666178  105.748184  102.838175   103.738176     6542800.0   
10  50.704382  103.514177  104.632181  101.694174   102.458174     8155160.0   
11  49.955130  102.084174  103.256179  100.420171   101.230172     7942280.0   
12  50.739348  101.230174  102.558177   99.940171   101.144173     8073040.0   
13  51.098988  100.918174  102.422177   99.608170   101.130172     8086640.0   

    Adj Close_roll5  
0               NaN  
1               NaN  
2               NaN  
3               NaN  
4         52.839251  
5         53.595495  
6         53.379711  
7         52.641448  
8         52.391698  
9         51.817271  
10        51.177909  
11        50.564522  
12        50.521565  
13        50.514572

> Lets see how we could repurpose **apply_rolling_mean** to use with different window sizes with an apply over columns...

In [55]:
def apply_rolling_mean(s, window=5):
    return s.rolling(window=window, center=False).mean()

_col_slice = ["Open", "High", "Low", "Close", "Volume", "Adj Close"]
_pre_bound = partial(apply_rolling_mean, window=3)
_rolling = _stock["GOOG"].loc[ : , _col_slice ].apply(_pre_bound)
_rolling.iloc[0:8]


Open        High         Low       Close        Volume  Adj Close
0         NaN         NaN         NaN         NaN           NaN        NaN
1         NaN         NaN         NaN         NaN           NaN        NaN
2  103.920178  108.873521  101.836841  106.016848  2.871897e+07  52.955468
3  107.666852  111.386857  104.373511  107.526848  1.886843e+07  53.709714
4  108.983520  111.026857  105.500180  106.756848  1.429820e+07  53.325099
5  107.050183  109.183522  104.036845  106.260181  1.056013e+07  53.077013
6  106.003515  108.190187  104.743513  106.686849  7.533967e+06  53.290135
7  106.110181  107.353519  104.120177  105.356845  6.197067e+06  52.625797

In [56]:
# We can also call rolling directly from the dataframe itself
_stock["GOOG"].loc[:,["Open"]].rolling(window=3, center=False).mean().head()

Open
0         NaN
1         NaN
2  103.920178
3  107.666852
4  108.983520

In [57]:
# this will be the same as the cells above
a = _stock["GOOG"].loc[:,["Open"]].rolling(window=3, center=False)
getattr(a, 'mean')().head()

Open
0         NaN
1         NaN
2  103.920178
3  107.666852
4  108.983520

In [58]:
# but now we can call other rolling stats programatically

stat_fns = ["mean", "sum"]

for itr in stat_fns:
    print("{} :\n".format(itr))
    print(getattr(a, itr)().head())
    print("\n---\n")

mean :

         Open
0         NaN
1         NaN
2  103.920178
3  107.666852
4  108.983520

---

sum :

         Open
0         NaN
1         NaN
2  311.760534
3  323.000555
4  326.950561

---



> We can bundle all this rolling logic in a function that rolls and merges the collumns that we want

In [59]:
roll_columns(_stock["GOOG"], "mean", column_slice=["Open"]).iloc[0:16]

rollingmean_10_Open
0                   NaN
1                   NaN
2                   NaN
3                   NaN
4                   NaN
5                   NaN
6                   NaN
7                   NaN
8                   NaN
9            105.129179
10           105.048180
11           105.042180
12           104.068178
13           103.018177
14           102.775176
15           102.440176

In [60]:
roll_columns(_stock["GOOG"], "mean", column_slice=["Open", "Close"]).iloc[0:16]

rollingmean_10_Open  rollingmean_10_Close
0                   NaN                   NaN
1                   NaN                   NaN
2                   NaN                   NaN
3                   NaN                   NaN
4                   NaN                   NaN
5                   NaN                   NaN
6                   NaN                   NaN
7                   NaN                   NaN
8                   NaN                   NaN
9            105.129179            104.761178
10           105.048180            104.878178
11           105.042180            104.048177
12           104.068178            103.266176
13           103.018177            103.009175
14           102.775176            102.640174
15           102.440176            102.382174

In [61]:
roll_columns(_stock["GOOG"], "mean", column_slice=["Open", "Close"], window=5).head()

rollingmean_5_Open  rollingmean_5_Close
0                 NaN                  NaN
1                 NaN                  NaN
2                 NaN                  NaN
3                 NaN                  NaN
4          105.592181           105.784181

In [62]:
roll_columns(_stock["GOOG"], "mean", column_slice=["Open", "Close"], window=3).head()

rollingmean_3_Open  rollingmean_3_Close
0                 NaN                  NaN
1                 NaN                  NaN
2          103.920178           106.016848
3          107.666852           107.526848
4          108.983520           106.756848

In [63]:
roll_columns(_stock["GOOG"], "mean", column_slice=["Open", "Close"], window=3, merge_result=True).head()

Date        Open        High         Low       Close    Volume  \
0 2004-08-19  100.000168  104.060182   95.960165  100.340176  44871300   
1 2004-08-20  101.010175  109.080187  100.500174  108.310183  22942800   
2 2004-08-23  110.750191  113.480193  109.050183  109.400185  18342800   
3 2004-08-24  111.240189  111.600192  103.570177  104.870176  15319700   
4 2004-08-25  104.960181  108.000187  103.880180  106.000184   9232100   

   Adj Close  rollingmean_3_Open  rollingmean_3_Close  
0  50.119968                 NaN                  NaN  
1  54.100990                 NaN                  NaN  
2  54.645447          103.920178           106.016848  
3  52.382705          107.666852           107.526848  
4  52.947145          108.983520           106.756848

> We really shouldn't leave the nan or other missing values there.

In [64]:
df_topad = roll_columns(_stock["GOOG"], "mean", column_slice=["Open", "Close"], window=3).head()
df_topad

rollingmean_3_Open  rollingmean_3_Close
0                 NaN                  NaN
1                 NaN                  NaN
2          103.920178           106.016848
3          107.666852           107.526848
4          108.983520           106.756848

In [65]:
df_topad.fillna(method="bfill")

rollingmean_3_Open  rollingmean_3_Close
0          103.920178           106.016848
1          103.920178           106.016848
2          103.920178           106.016848
3          107.666852           107.526848
4          108.983520           106.756848

> we could acomodate this on our function....

In [66]:
roll_columns(_stock["GOOG"], "mean", column_slice=["Open", "Close"], window=3, pad_result=False).head()

rollingmean_3_Open  rollingmean_3_Close
0                 NaN                  NaN
1                 NaN                  NaN
2          103.920178           106.016848
3          107.666852           107.526848
4          108.983520           106.756848

In [67]:
roll_columns(_stock["GOOG"], "mean", column_slice=["Open", "Close"], window=3, pad_result=True).head()

rollingmean_3_Open  rollingmean_3_Close
0          103.920178           106.016848
1          103.920178           106.016848
2          103.920178           106.016848
3          107.666852           107.526848
4          108.983520           106.756848

We can look at how to compute the daily returns by shifting our data

In [68]:
_sample_goog_close = _stock["GOOG"].loc[:,["Close"]].iloc[:15]
_sample_goog_close

Close
0   100.340176
1   108.310183
2   109.400185
3   104.870176
4   106.000184
5   107.910182
6   106.150181
7   102.010172
8   102.370175
9   100.250171
10  101.510173
11  100.010169
12  101.580175
13  102.300173
14  102.310174

In [69]:
%%timeit

_rr = (_sample_goog_close[1:] / _sample_goog_close[:-1].values) - 1
_rr.ix[0] = 0
_rr



100 loops, best of 3: 4.18 ms per loop


In [70]:
%%timeit

# this code does the same as the cell above, but as we can see it is much faster
_rr = (_sample_goog_close / _sample_goog_close.shift(1)) - 1
_rr.ix[0] = 0
_rr

1000 loops, best of 3: 1.24 ms per loop


> Lets try with multiple collumns

In [71]:

_rr = (_sample_goog_close / _sample_goog_close.shift(1)) - 1
_rr.ix[0] = 0
_rr

Close
0   0.000000
1   0.079430
2   0.010064
3  -0.041408
4   0.010775
5   0.018019
6  -0.016310
7  -0.039001
8   0.003529
9  -0.020709
10  0.012569
11 -0.014777
12  0.015698
13  0.007088
14  0.000098

In [72]:
_rr = (_sample_goog_close / _sample_goog_close.shift(1)) - 1
_rr = _rr.fillna(0)
_rr

Close
0   0.000000
1   0.079430
2   0.010064
3  -0.041408
4   0.010775
5   0.018019
6  -0.016310
7  -0.039001
8   0.003529
9  -0.020709
10  0.012569
11 -0.014777
12  0.015698
13  0.007088
14  0.000098

> Lets try out some functions of our own. Built on top of this

In [145]:
import numpy as np
df = pd.DataFrame([[1, 1.2, 1.8],[1, 0.8, 1.6],[1,-0.5,0.1]]).T
df

0    1    2
0  1.0  1.0  1.0
1  1.2  0.8 -0.5
2  1.8  1.6  0.1

In [74]:
timewindow_diff(df)

timewindow_diff_1_0  timewindow_diff_1_1  timewindow_diff_1_2
0                  NaN                  NaN                  NaN
1                  0.2                 -0.2                 -1.5
2                  0.6                  0.8                  0.6

In [75]:
timewindow_diff(df, shift_window=2, fillna=True)

timewindow_diff_2_0  timewindow_diff_2_1  timewindow_diff_2_2
0                  0.0                  0.0                  0.0
1                  0.0                  0.0                  0.0
2                  0.8                  0.6                 -0.9

In [76]:
timewindow_diff(df, column_slice=[1], fillna=True, merge_result=True)

0    1    2  timewindow_diff_1_1
0  1.0  1.0  1.0                  0.0
1  1.2  0.8 -0.5                 -0.2
2  1.8  1.6  0.1                  0.8

In [77]:
df

0    1    2
0  1.0  1.0  1.0
1  1.2  0.8 -0.5
2  1.8  1.6  0.1

In [78]:
timewindow_return(df)

timewindow_return_1_0  timewindow_return_1_1  timewindow_return_1_2
0                    NaN                    NaN                    NaN
1                    0.2                   -0.2                   -1.5
2                    0.5                    1.0                   -1.2

> timewindow_return gives us the variation from an initial value. See the explanation below.

In [79]:
n = 200
p = 1.5
v = p * n
print("You buy {} shares at {}. Your value is {}".format(n,p,v))
np = 1.0
nv = np * n
r = ((np / p) - 1.0)
print("The price changes to {}. You new value is {}.".format(np, nv))

print ("Your return is {} ({} of your initial value)".format(v*r, r))

You buy 200 shares at 1.5. Your value is 300.0
The price changes to 1.0. You new value is 200.0.
Your return is -100.00000000000001 (-0.33333333333333337 of your initial value)


In [80]:
df

0    1    2
0  1.0  1.0  1.0
1  1.2  0.8 -0.5
2  1.8  1.6  0.1

In [81]:
timewindow_cumdiff(df)

timewindow_cumdiff_0  timewindow_cumdiff_1  timewindow_cumdiff_2
0                   0.0                   0.0                   0.0
1                   0.2                  -0.2                  -1.5
2                   0.8                   0.6                  -0.9

In [82]:
timewindow_cumreturn(df)

timewindow_cumreturn_0  timewindow_cumreturn_1  timewindow_cumreturn_2
0                     0.0                     0.0                     0.0
1                     0.2                    -0.2                    -1.5
2                     0.8                     0.6                    -0.9

> Now, lets take a real look at dates and how to manage them

In [85]:
import numpy as np

# This can create an array with a date for us
ledate = np.array('2015-12-30', dtype=np.datetime64)
ledate

array(datetime.date(2015, 12, 30), dtype='datetime64[D]')

> we can obviously do vectorized operation on *ledate*

In [87]:
lerange_12day = ledate + np.arange(12)
lerange_12day

array(['2015-12-30', '2015-12-31', '2016-01-01', '2016-01-02',
       '2016-01-03', '2016-01-04', '2016-01-05', '2016-01-06',
       '2016-01-07', '2016-01-08', '2016-01-09', '2016-01-10'], dtype='datetime64[D]')

> We can also use *np.datetime64* in order to create a date constant. It's possible to give it a time format

In [88]:
print(np.datetime64('2015-07-04 12:00'))
print(np.datetime64('2015-07-04 12:59:59.50', 'ns'))

2015-07-04T12:00
2015-07-04T12:59:59.500000000


> These are the code formats possible with numpy

|Code    | Meaning     | Time span (relative) | Time span (absolute)   |
|--------|-------------|----------------------|------------------------|
| ``Y``  | Year	       | ± 9.2e18 years       | [9.2e18 BC, 9.2e18 AD] |
| ``M``  | Month       | ± 7.6e17 years       | [7.6e17 BC, 7.6e17 AD] |
| ``W``  | Week	       | ± 1.7e17 years       | [1.7e17 BC, 1.7e17 AD] |
| ``D``  | Day         | ± 2.5e16 years       | [2.5e16 BC, 2.5e16 AD] |
| ``h``  | Hour        | ± 1.0e15 years       | [1.0e15 BC, 1.0e15 AD] |
| ``m``  | Minute      | ± 1.7e13 years       | [1.7e13 BC, 1.7e13 AD] |
| ``s``  | Second      | ± 2.9e12 years       | [ 2.9e9 BC, 2.9e9 AD]  |
| ``ms`` | Millisecond | ± 2.9e9 years        | [ 2.9e6 BC, 2.9e6 AD]  |
| ``us`` | Microsecond | ± 2.9e6 years        | [290301 BC, 294241 AD] |
| ``ns`` | Nanosecond  | ± 292 years          | [ 1678 AD, 2262 AD]    |
| ``ps`` | Picosecond  | ± 106 days           | [ 1969 AD, 1970 AD]    |
| ``fs`` | Femtosecond | ± 2.6 hours          | [ 1969 AD, 1970 AD]    |
| ``as`` | Attosecond  | ± 9.2 seconds        | [ 1969 AD, 1970 AD]    |

> We can also use pandas to manipulate our dates

In [89]:
ledate = pd.to_datetime("4th of July, 2015")
ledate

Timestamp('2015-07-04 00:00:00')

In [90]:
ledate.strftime('%A')

'Saturday'

> We can also use vectorized operations 

In [91]:
lerange_12days = ledate + pd.to_timedelta(np.arange(12), 'D')
lerange_12days

DatetimeIndex(['2015-07-04', '2015-07-05', '2015-07-06', '2015-07-07',
               '2015-07-08', '2015-07-09', '2015-07-10', '2015-07-11',
               '2015-07-12', '2015-07-13', '2015-07-14', '2015-07-15'],
              dtype='datetime64[ns]', freq=None)

In [92]:
lerange_12months = ledate + pd.to_timedelta(np.arange(12), 'M')
lerange_12months

DatetimeIndex(['2015-07-04 00:00:00', '2015-08-03 10:29:06',
               '2015-09-02 20:58:12', '2015-10-03 07:27:18',
               '2015-11-02 17:56:24', '2015-12-03 04:25:30',
               '2016-01-02 14:54:36', '2016-02-02 01:23:42',
               '2016-03-03 11:52:48', '2016-04-02 22:21:54',
               '2016-05-03 08:51:00', '2016-06-02 19:20:06'],
              dtype='datetime64[ns]', freq=None)

In [93]:
lerange_12minutes = ledate + pd.to_timedelta(np.arange(12), 'm')
lerange_12minutes

DatetimeIndex(['2015-07-04 00:00:00', '2015-07-04 00:01:00',
               '2015-07-04 00:02:00', '2015-07-04 00:03:00',
               '2015-07-04 00:04:00', '2015-07-04 00:05:00',
               '2015-07-04 00:06:00', '2015-07-04 00:07:00',
               '2015-07-04 00:08:00', '2015-07-04 00:09:00',
               '2015-07-04 00:10:00', '2015-07-04 00:11:00'],
              dtype='datetime64[ns]', freq=None)

> We can also use *pd.date_range* to achieve this

In [94]:
pd.date_range('2015-07-03', '2015-07-10')

DatetimeIndex(['2015-07-03', '2015-07-04', '2015-07-05', '2015-07-06',
               '2015-07-07', '2015-07-08', '2015-07-09', '2015-07-10'],
              dtype='datetime64[ns]', freq='D')

In [95]:
pd.date_range('2015-07-03', periods=4)

DatetimeIndex(['2015-07-03', '2015-07-04', '2015-07-05', '2015-07-06'], dtype='datetime64[ns]', freq='D')

In [96]:
pd.date_range('2015-07-03', periods=8, freq='H')

DatetimeIndex(['2015-07-03 00:00:00', '2015-07-03 01:00:00',
               '2015-07-03 02:00:00', '2015-07-03 03:00:00',
               '2015-07-03 04:00:00', '2015-07-03 05:00:00',
               '2015-07-03 06:00:00', '2015-07-03 07:00:00'],
              dtype='datetime64[ns]', freq='H')

In [97]:
pd.period_range('2015-07', periods=8, freq='M')

PeriodIndex(['2015-07', '2015-08', '2015-09', '2015-10', '2015-11', '2015-12',
             '2016-01', '2016-02'],
            dtype='period[M]', freq='M')

In [98]:
pd.timedelta_range(0, periods=10, freq='H')

TimedeltaIndex(['00:00:00', '01:00:00', '02:00:00', '03:00:00', '04:00:00',
                '05:00:00', '06:00:00', '07:00:00', '08:00:00', '09:00:00'],
               dtype='timedelta64[ns]', freq='H')

#### Frequencies and Offsets
(Taken from [Python Data Science Handbook](http://shop.oreilly.com/product/0636920034919.do) by Jake VanderPlas; the content is available [on GitHub](https://github.com/jakevdp/PythonDataScienceHandbook))

Fundamental to these Pandas time series tools is the concept of a frequency or date offset.
Just as we saw the ``D`` (day) and ``H`` (hour) codes above, we can use such codes to specify any desired frequency spacing.
The following table summarizes the main codes available:

| Code   | Description         | Code   | Description          |
|--------|---------------------|--------|----------------------|
| ``D``  | Calendar day        | ``B``  | Business day         |
| ``W``  | Weekly              |        |                      |
| ``M``  | Month end           | ``BM`` | Business month end   |
| ``Q``  | Quarter end         | ``BQ`` | Business quarter end |
| ``A``  | Year end            | ``BA`` | Business year end    |
| ``H``  | Hours               | ``BH`` | Business hours       |
| ``T``  | Minutes             |        |                      |
| ``S``  | Seconds             |        |                      |
| ``L``  | Milliseonds         |        |                      |
| ``U``  | Microseconds        |        |                      |
| ``N``  | nanoseconds         |        |                      |

| Code    | Description            || Code    | Description            |
|---------|------------------------||---------|------------------------|
| ``MS``  | Month start            ||``BMS``  | Business month start   |
| ``QS``  | Quarter start          ||``BQS``  | Business quarter start |
| ``AS``  | Year start             ||``BAS``  | Business year start    |

In [99]:
## This will give us 9 spans of 2 and a half hours each
pd.timedelta_range(0, periods=9, freq="2H30T")

TimedeltaIndex(['00:00:00', '02:30:00', '05:00:00', '07:30:00', '10:00:00',
                '12:30:00', '15:00:00', '17:30:00', '20:00:00'],
               dtype='timedelta64[ns]', freq='150T')

In [100]:
aapl = _stock["AAPL"]
aapl.head()

Date       Open       High     Low   Close     Volume  Adj Close
0 1980-12-12  28.750000  28.875000  28.750  28.750  117258400   0.426842
1 1980-12-15  27.375001  27.375001  27.250  27.250   43971200   0.404572
2 1980-12-16  25.375000  25.375000  25.250  25.250   26432000   0.374879
3 1980-12-17  25.875000  25.999999  25.875  25.875   21610400   0.384158
4 1980-12-18  26.625000  26.750000  26.625  26.625   18362400   0.395293

In [101]:
aapl.corr()

Open      High       Low     Close    Volume  Adj Close
Open       1.000000  0.999929  0.999902  0.999826  0.201803   0.615601
High       0.999929  1.000000  0.999882  0.999921  0.204513   0.615001
Low        0.999902  0.999882  1.000000  0.999915  0.197761   0.616441
Close      0.999826  0.999921  0.999915  1.000000  0.201303   0.615769
Volume     0.201803  0.204513  0.197761  0.201303  1.000000   0.023650
Adj Close  0.615601  0.615001  0.616441  0.615769  0.023650   1.000000

In [102]:
goog = _stock["GOOG"]

In [103]:
_rolling = aapl.loc[:100, ["Close"]].rolling(window=10, center=True)

_rolling.corr(other=goog["Open"]).iloc[:16]

Close
0        NaN
1        NaN
2        NaN
3        NaN
4        NaN
5  -0.492011
6  -0.696666
7  -0.874905
8  -0.850180
9  -0.807468
10 -0.850498
11 -0.855884
12 -0.811677
13 -0.844038
14 -0.784980
15 -0.783991

In [104]:
aapl["Close"].corr(goog["Close"])

0.48850580664079235

> Lets take a look at the exchanges

In [105]:
_tkrs = ["^IXIC", "^GSPC"]

_r = initial_dataload(_tkrs)

_exchanges = dict()

for t in _tkrs:
    _exchanges[t] = pd.read_csv("{}/{}.csv".format(DATA_PATH, t))

(1/2) Recv. and Stored ^IXIC
(2/2) Recv. and Stored ^GSPC
Took 0:00:03.362369


In [116]:
df_beta = pd.DataFrame()

df_beta["IXIC"] = _exchanges["^IXIC"].iloc[-100:, :]["Close"]
df_beta["GOOG"] = _stock["GOOG"].iloc[-100:, :]["Close"]

In [121]:
def calc_beta(df):
	# first column is the market
	X = df.values[:, [0]]
	# prepend a column of ones for the intercept
	X = np.concatenate([np.ones_like(X), X], axis=1)
	# matrix algebra
	b = np.linalg.pinv(X.T.dot(X)).dot(X.T).dot(df.values[:, 1:])
	return pd.Series(b[1], df.columns[1:], name='Beta')




In [126]:
df_beta.head()

IXIC  GOOG
11493  5213.220215   NaN
11494  5227.209961   NaN
11495  5249.899902   NaN
11496  5275.910156   NaN
11497  5283.930176   NaN

In [128]:
_stock["GOOG"].iloc[-100:, :]["Close"]

3030    767.049988
3031    768.780029
3032    771.460022
3033    780.080017
3034    780.349976
3035    775.320007
3036    759.659973
3037    769.020020
3038    759.690002
3039    762.489990
3040    771.760010
3041    768.880005
3042    765.700012
3043    771.409973
3044    776.219971
3045    787.210022
3046    786.900024
3047    774.210022
3048    783.010010
3049    781.559998
3050    775.010010
3051    777.289978
3052    772.559998
3053    776.429993
3054    776.469971
3055    776.859985
3056    775.080017
3057    785.940002
3058    783.070007
3059    786.140015
           ...    
3100    789.289978
3101    789.270020
3102    796.099976
3103    797.070007
3104    797.849976
3105    790.799988
3106    794.200012
3107    796.419983
3108    794.559998
3109    791.260010
3110    789.909973
3111    791.549988
3112    785.049988
3113    782.789978
3114    771.820007
3115    786.140015
3116    786.900024
3117    794.020020
3118    806.150024
3119    806.650024
3120    804.789978
3121    807.

In [129]:
dates = pd.date_range('1995-12-31', periods=480, freq='M', name='Date')
stoks = pd.Index(['s{:04d}'.format(i) for i in range(4000)])
df = pd.DataFrame(np.random.rand(480, 4000), dates, stoks)

In [130]:
df.iloc[:5, :5]


s0000     s0001     s0002     s0003     s0004
Date                                                        
1995-12-31  0.458905  0.411290  0.795909  0.634213  0.192437
1996-01-31  0.974046  0.771754  0.380362  0.002019  0.837676
1996-02-29  0.310043  0.548079  0.647215  0.726447  0.020373
1996-03-31  0.241670  0.936383  0.129207  0.681234  0.336017
1996-04-30  0.254399  0.270134  0.509911  0.957042  0.096045

In [131]:
def roll(df, w):
    # stack df.values w-times shifted once at each stack
    roll_array = np.dstack([df.values[i:i+w, :] for i in range(len(df.index) - w + 1)]).T
    # roll_array is now a 3-D array and can be read into
    # a pandas panel object
    panel = pd.Panel(roll_array, 
                     items=df.index[w-1:],
                     major_axis=df.columns,
                     minor_axis=pd.Index(range(w), name='roll'))
    # convert to dataframe and pivot + groupby
    # is now ready for any action normally performed
    # on a groupby object
    return panel.to_frame().unstack().T.groupby(level=0)


In [132]:
def beta(df):
    # first column is the market
    X = df.values[:, [0]]
    # prepend a column of ones for the intercept
    X = np.concatenate([np.ones_like(X), X], axis=1)
    # matrix algebra
    b = np.linalg.pinv(X.T.dot(X)).dot(X.T).dot(df.values[:, 1:])
    return pd.Series(b[1], df.columns[1:], name='Beta')


In [133]:
rdf = roll(df, 12)
betas = rdf.apply(beta)


In [134]:
def calc_beta(df):
    np_array = df.values
    m = np_array[:,0] # market returns are column zero from numpy array
    s = np_array[:,1] # stock returns are column one from numpy array
    covariance = np.cov(s,m) # Calculate covariance between stock and market
    beta = covariance[0,1]/covariance[1,1]
    return beta


In [135]:
print(calc_beta(df.iloc[:12, :2]))


0.00886393088116


In [136]:
print(beta(df.iloc[:12, :2]))


s0001    0.008864
Name: Beta, dtype: float64


In [137]:
betas = rdf.apply(beta)
betas.iloc[:5, :5]


major          s0001     s0002     s0003     s0004     s0005
Date                                                        
1996-11-30  0.008864  0.231190 -0.512688  0.088701  0.112922
1996-12-31  0.064562  0.336461 -0.433014 -0.121619  0.198959
1997-01-31 -0.169694  0.262055 -0.158651 -0.447662  0.178437
1997-02-28 -0.154411  0.311005 -0.148477 -0.509510  0.198030
1997-03-31 -0.135591  0.317450 -0.093781 -0.445259  0.145199

In [139]:
%%timeit 
roll(df, 12).apply(beta)

1 loop, best of 3: 4.77 s per loop


In [1]:
from scipy import stats
import numpy as np
x = np.random.random(10)
y = np.random.random(10)
slope, intercept, r_value, p_value, std_err = stats.linregress(x,y)


In [10]:
_stock = dict()
_stock["TWTR"] = pd.read_csv("{}/{}.csv".format(DATA_PATH, "TWTR"))

In [8]:
timespan = None

timespan = {
    "short_term": [1,5,10]
    ,"medium_term": [30,50,70]
    ,"long_term": [100, 200, 400]
}


In [9]:
gains = None

gains = {
    "low": [0.01]
    ,"medium": [0.05, 0.07, 0.1]
    ,"high": [0.2, 0.4, 1.0]
}

In [10]:
losses = None

losses = {
    "low": [-0.05, -0.1]
    ,"medium": [-0.3, -0.5]
    ,"high": [-0.8, -1.0, -2.0]
}

In [36]:
itr_df = _stock["TWTR"];

itr_df.head()

Date       Open       High        Low      Close     Volume  \
0  2013-11-07  45.099998  50.090000  44.000000  44.900002  117701600   
1  2013-11-08  45.930000  46.939999  40.689999  41.650002   27925300   
2  2013-11-11  40.500000  43.000000  39.400002  42.900002   16113900   
3  2013-11-12  43.660000  43.779999  41.830002  41.900002    6316700   
4  2013-11-13  41.029999  42.869999  40.759998  42.599998    8688300   

   Adj Close  
0  44.900002  
1  41.650002  
2  42.900002  
3  41.900002  
4  42.599998

In [121]:
## calculate diff moves for timespans
## calculate returns for timespans
## calculate moving averages
## calculate bollinger bands

In [122]:
## calculate diff moves for timespans

itr_df = _stock["TWTR"];

for ts_name in timespan:
    for t in timespan[ts_name]:
        itr_df = timewindow_diff(itr_df, column_slice=["Close", "Volume"], shift_window=t, merge_result=True)        

itr_df.head()


Date       Open       High        Low      Close     Volume  \
0  2013-11-07  45.099998  50.090000  44.000000  44.900002  117701600   
1  2013-11-08  45.930000  46.939999  40.689999  41.650002   27925300   
2  2013-11-11  40.500000  43.000000  39.400002  42.900002   16113900   
3  2013-11-12  43.660000  43.779999  41.830002  41.900002    6316700   
4  2013-11-13  41.029999  42.869999  40.759998  42.599998    8688300   

   Adj Close  timewindow_diff_100_Close  timewindow_diff_100_Volume  \
0  44.900002                        NaN                         NaN   
1  41.650002                        NaN                         NaN   
2  42.900002                        NaN                         NaN   
3  41.900002                        NaN                         NaN   
4  42.599998                        NaN                         NaN   

   timewindow_diff_200_Close            ...              \
0                        NaN            ...               
1                        NaN            ...               
2                        NaN            ...               
3                        NaN            ...               
4                        NaN            ...               

   timewindow_diff_50_Close  timewindow_diff_50_Volume  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

   timewindow_diff_70_Close  timewindow_diff_70_Volume  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

   timewindow_diff_1_Close  timewindow_diff_1_Volume  timewindow_diff_5_Close  \
0                      NaN                       NaN                      NaN   
1                -3.250000               -89776300.0                      NaN   
2                 1.250000               -11811400.0                      NaN   
3                -1.000000                -9797200.0                      NaN   
4                 0.699996                 2371600.0                      NaN   

   timewindow_diff_5_Volume  timewindow_diff_10_Close  \
0                       NaN                       NaN   
1                       NaN                       NaN   
2                       NaN                       NaN   
3                       NaN                       NaN   
4                       NaN                       NaN   

   timewindow_diff_10_Volume  
0                        NaN  
1                        NaN  
2                        NaN  
3                        NaN  
4                        NaN  

[5 rows x 25 columns]

In [27]:
## calculate returns for timespans

itr_df = _stock["TWTR"];

for ts_name in timespan:
    for t in timespan[ts_name]:
        itr_df = timewindow_return(itr_df, column_slice=["Close", "Volume"], shift_window=t, merge_result=True)        

itr_df.head()



NameError: name '_stock' is not defined

In [124]:
## calculate moving averages

itr_df = _stock["TWTR"];

for ts_name in timespan:
    for t in timespan[ts_name]:
        itr_df = roll_columns(itr_df, "mean", column_slice=["Close", "Volume"], window=t, merge_result=True)        

itr_df.head()


Date       Open       High        Low      Close     Volume  \
0  2013-11-07  45.099998  50.090000  44.000000  44.900002  117701600   
1  2013-11-08  45.930000  46.939999  40.689999  41.650002   27925300   
2  2013-11-11  40.500000  43.000000  39.400002  42.900002   16113900   
3  2013-11-12  43.660000  43.779999  41.830002  41.900002    6316700   
4  2013-11-13  41.029999  42.869999  40.759998  42.599998    8688300   

   Adj Close  Close_roll_mean_100  Volume_roll_mean_100  Close_roll_mean_200  \
0  44.900002                  NaN                   NaN                  NaN   
1  41.650002                  NaN                   NaN                  NaN   
2  42.900002                  NaN                   NaN                  NaN   
3  41.900002                  NaN                   NaN                  NaN   
4  42.599998                  NaN                   NaN                  NaN   

          ...           Close_roll_mean_50  Volume_roll_mean_50  \
0         ...                          NaN                  NaN   
1         ...                          NaN                  NaN   
2         ...                          NaN                  NaN   
3         ...                          NaN                  NaN   
4         ...                          NaN                  NaN   

   Close_roll_mean_70  Volume_roll_mean_70  Close_roll_mean_1  \
0                 NaN                  NaN          44.900002   
1                 NaN                  NaN          41.650002   
2                 NaN                  NaN          42.900002   
3                 NaN                  NaN          41.900002   
4                 NaN                  NaN          42.599998   

   Volume_roll_mean_1  Close_roll_mean_5  Volume_roll_mean_5  \
0         117701600.0                NaN                 NaN   
1          27925300.0                NaN                 NaN   
2          16113900.0                NaN                 NaN   
3           6316700.0                NaN                 NaN   
4           8688300.0          42.790001          35349160.0   

   Close_roll_mean_10  Volume_roll_mean_10  
0                 NaN                  NaN  
1                 NaN                  NaN  
2                 NaN                  NaN  
3                 NaN                  NaN  
4                 NaN                  NaN  

[5 rows x 25 columns]

In [186]:
## calculate bollinger bands

itr_df = pd.DataFrame()
res_df = pd.DataFrame()

for ts_name in timespan:
    if ts_name in ["medium_term", "long_term"]:
        for t in timespan[ts_name]:
            itr_df = roll_columns(_stock["TWTR"], "std", column_slice=["Close"], window=t, merge_result=False, scaler=2, pad_result=True)
            upper_band = itr_df.apply(lambda x: s + x)
            lower_band = itr_df.apply(lambda x: s - x)
            
            roll_name = "{}_roll_2std_{}".format("Close", t)
            upper_name = "{}_bollinger_{}_up".format("Close", t)
            lower_name = "{}_bollinger_{}_low".format("Close", t)
            
            res_df[roll_name] = itr_df.iloc[:, 0]
            res_df[upper_name] = upper_band
            res_df[lower_name] = lower_band

            
res_df.head()

Close_roll_2std_100  Close_bollinger_100_up  Close_bollinger_100_low  \
0            16.146695               61.046697                28.753307   
1            16.146695               57.796697                25.503307   
2            16.146695               59.046697                26.753307   
3            16.146695               58.046697                25.753307   
4            16.146695               58.746693                26.453303   

   Close_roll_2std_200  Close_bollinger_200_up  Close_bollinger_200_low  \
0            19.768411               64.668413                25.131591   
1            19.768411               61.418413                21.881591   
2            19.768411               62.668413                23.131591   
3            19.768411               61.668413                22.131591   
4            19.768411               62.368409                22.831587   

   Close_roll_2std_400  Close_bollinger_400_up  Close_bollinger_400_low  \
0            16.612795               61.512797                28.287207   
1            16.612795               58.262797                25.037207   
2            16.612795               59.512797                26.287207   
3            16.612795               58.512797                25.287207   
4            16.612795               59.212793                25.987203   

   Close_roll_2std_30  Close_bollinger_30_up  Close_bollinger_30_low  \
0           12.537888              57.437890               32.362114   
1           12.537888              54.187890               29.112114   
2           12.537888              55.437890               30.362114   
3           12.537888              54.437890               29.362114   
4           12.537888              55.137886               30.062110   

   Close_roll_2std_50  Close_bollinger_50_up  Close_bollinger_50_low  \
0           19.990288              64.890290               24.909714   
1           19.990288              61.640290               21.659714   
2           19.990288              62.890290               22.909714   
3           19.990288              61.890290               21.909714   
4           19.990288              62.590286               22.609710   

   Close_roll_2std_70  Close_bollinger_70_up  Close_bollinger_70_low  
0           18.532434              63.432436               26.367568  
1           18.532434              60.182436               23.117568  
2           18.532434              61.432436               24.367568  
3           18.532434              60.432436               23.367568  
4           18.532434              61.132432               24.067564

In [175]:
_stock = dict()
_stock["TWTR"] = pd.read_csv("{}/{}.csv".format(DATA_PATH, "TWTR"))

In [179]:
_r = load_raw_frame("NDX")
_r.head()

Date    Open    High     Low   Close  Volume  Adj Close
0  1985-01-31  125.00  125.00  125.00  125.00       0      89.58
1  1985-02-28  123.52  123.52  123.52  123.52       0      88.52
2  1985-03-25  115.02  115.02  115.02  115.02       0      82.43
3  1985-03-26  115.25  115.25  115.25  115.25       0      82.59
4  1985-03-27  116.38  116.38  116.38  116.38       0      83.40

In [92]:
aapl = load_raw_frame("AAPL")
aapl.head()

Date       Open       High     Low   Close     Volume  Adj Close
0  1980-12-12  28.750000  28.875000  28.750  28.750  117258400   0.425000
1  1980-12-15  27.375001  27.375001  27.250  27.250   43971200   0.402826
2  1980-12-16  25.375000  25.375000  25.250  25.250   26432000   0.373261
3  1980-12-17  25.875000  25.999999  25.875  25.875   21610400   0.382500
4  1980-12-18  26.625000  26.750000  26.625  26.625   18362400   0.393587

In [93]:
timespan = None

timespan = {
    "medium_term": [4,7]
}


In [94]:
aapl = load_raw_frame("AAPL")

In [95]:
calc_diff_moves(aapl, timespan=timespan, column_slice=["Close", "Volume"], fillna=True, merge_result=True)

Date        Open        High         Low       Close     Volume  \
0     1980-12-12   28.750000   28.875000   28.750000   28.750000  117258400   
1     1980-12-15   27.375001   27.375001   27.250000   27.250000   43971200   
2     1980-12-16   25.375000   25.375000   25.250000   25.250000   26432000   
3     1980-12-17   25.875000   25.999999   25.875000   25.875000   21610400   
4     1980-12-18   26.625000   26.750000   26.625000   26.625000   18362400   
5     1980-12-19   28.249999   28.374999   28.249999   28.249999   12157600   
6     1980-12-22   29.625000   29.750000   29.625000   29.625000    9340800   
7     1980-12-23   30.874999   30.999999   30.874999   30.874999   11737600   
8     1980-12-24   32.500000   32.624999   32.500000   32.500000   12000800   
9     1980-12-26   35.500001   35.625001   35.500001   35.500001   13893600   
10    1980-12-29   36.000000   36.124999   36.000000   36.000000   23290400   
11    1980-12-30   35.249999   35.249999   35.125000   35.125000   17220000   
12    1980-12-31   34.250000   34.250000   34.125000   34.125000    8937600   
13    1981-01-02   34.499999   34.750001   34.499999   34.499999    5415200   
14    1981-01-05   33.875001   33.875001   33.750001   33.750001    8932000   
15    1981-01-06   32.375000   32.375000   32.250000   32.250000   11289600   
16    1981-01-07   30.999999   30.999999   30.874999   30.874999   13921600   
17    1981-01-08   30.375001   30.375001   30.250001   30.250001    9956800   
18    1981-01-09   31.874999   32.000001   31.874999   31.874999    5376000   
19    1981-01-12   31.874999   31.874999   31.625000   31.625000    5924800   
20    1981-01-13   30.625000   30.625000   30.500000   30.500000    5762400   
21    1981-01-14   30.625000   30.750000   30.625000   30.625000    3572800   
22    1981-01-15   31.250001   31.500000   31.250001   31.250001    3516800   
23    1981-01-16   31.125001   31.125001   30.999999   30.999999    3348800   
24    1981-01-19   32.875001   33.000001   32.875001   32.875001   10393600   
25    1981-01-20   32.000001   32.000001   31.874999   31.874999    7520800   
26    1981-01-21   32.500000   32.749999   32.500000   32.500000    3976000   
27    1981-01-22   32.875001   33.125000   32.875001   32.875001    8887200   
28    1981-01-23   32.875001   33.000001   32.749999   32.749999    2805600   
29    1981-01-26   32.375000   32.375000   32.250000   32.250000    6160000   
...          ...         ...         ...         ...         ...        ...   
9090  2016-12-29  116.449997  117.110001  116.400002  116.730003   15039500   
9091  2016-12-30  116.650002  117.199997  115.430000  115.820000   30586300   
9092  2017-01-03  115.800003  116.330002  114.760002  116.150002   28781900   
9093  2017-01-04  115.849998  116.510002  115.750000  116.019997   21118100   
9094  2017-01-05  115.919998  116.860001  115.809998  116.610001   22193600   
9095  2017-01-06  116.779999  118.160004  116.470001  117.910004   31751900   
9096  2017-01-09  117.949997  119.430000  117.940002  118.989998   33387600   
9097  2017-01-10  118.769997  119.379997  118.300003  119.110001   24462100   
9098  2017-01-11  118.739998  119.930000  118.599998  119.750000   27588600   
9099  2017-01-12  118.900002  119.300003  118.209999  119.250000   27086200   
9100  2017-01-13  119.110001  119.620003  118.809998  119.040001   26111900   
9101  2017-01-17  118.339996  120.239998  118.220001  120.000000   34439800   
9102  2017-01-18  120.000000  120.500000  119.709999  119.989998   23713000   
9103  2017-01-19  119.400002  120.089996  119.370003  119.779999   25597300   
9104  2017-01-20  120.449997  120.449997  119.730003  120.000000   32597900   
9105  2017-01-23  120.000000  120.809998  119.769997  120.080002   22050200   
9106  2017-01-24  119.550003  120.099998  119.500000  119.970001   23211000   
9107  2017-01-25  120.419998  122.099998  120.279999  121.879997   32377600   
9108  2017-01-26  121.669998  122.440002  121.599998  121.9400

In [96]:
calc_return(aapl, timespan=timespan, column_slice=["Close", "Volume"], merge_result=True)

Date        Open        High         Low       Close     Volume  \
0     1980-12-12   28.750000   28.875000   28.750000   28.750000  117258400   
1     1980-12-15   27.375001   27.375001   27.250000   27.250000   43971200   
2     1980-12-16   25.375000   25.375000   25.250000   25.250000   26432000   
3     1980-12-17   25.875000   25.999999   25.875000   25.875000   21610400   
4     1980-12-18   26.625000   26.750000   26.625000   26.625000   18362400   
5     1980-12-19   28.249999   28.374999   28.249999   28.249999   12157600   
6     1980-12-22   29.625000   29.750000   29.625000   29.625000    9340800   
7     1980-12-23   30.874999   30.999999   30.874999   30.874999   11737600   
8     1980-12-24   32.500000   32.624999   32.500000   32.500000   12000800   
9     1980-12-26   35.500001   35.625001   35.500001   35.500001   13893600   
10    1980-12-29   36.000000   36.124999   36.000000   36.000000   23290400   
11    1980-12-30   35.249999   35.249999   35.125000   35.125000   17220000   
12    1980-12-31   34.250000   34.250000   34.125000   34.125000    8937600   
13    1981-01-02   34.499999   34.750001   34.499999   34.499999    5415200   
14    1981-01-05   33.875001   33.875001   33.750001   33.750001    8932000   
15    1981-01-06   32.375000   32.375000   32.250000   32.250000   11289600   
16    1981-01-07   30.999999   30.999999   30.874999   30.874999   13921600   
17    1981-01-08   30.375001   30.375001   30.250001   30.250001    9956800   
18    1981-01-09   31.874999   32.000001   31.874999   31.874999    5376000   
19    1981-01-12   31.874999   31.874999   31.625000   31.625000    5924800   
20    1981-01-13   30.625000   30.625000   30.500000   30.500000    5762400   
21    1981-01-14   30.625000   30.750000   30.625000   30.625000    3572800   
22    1981-01-15   31.250001   31.500000   31.250001   31.250001    3516800   
23    1981-01-16   31.125001   31.125001   30.999999   30.999999    3348800   
24    1981-01-19   32.875001   33.000001   32.875001   32.875001   10393600   
25    1981-01-20   32.000001   32.000001   31.874999   31.874999    7520800   
26    1981-01-21   32.500000   32.749999   32.500000   32.500000    3976000   
27    1981-01-22   32.875001   33.125000   32.875001   32.875001    8887200   
28    1981-01-23   32.875001   33.000001   32.749999   32.749999    2805600   
29    1981-01-26   32.375000   32.375000   32.250000   32.250000    6160000   
...          ...         ...         ...         ...         ...        ...   
9090  2016-12-29  116.449997  117.110001  116.400002  116.730003   15039500   
9091  2016-12-30  116.650002  117.199997  115.430000  115.820000   30586300   
9092  2017-01-03  115.800003  116.330002  114.760002  116.150002   28781900   
9093  2017-01-04  115.849998  116.510002  115.750000  116.019997   21118100   
9094  2017-01-05  115.919998  116.860001  115.809998  116.610001   22193600   
9095  2017-01-06  116.779999  118.160004  116.470001  117.910004   31751900   
9096  2017-01-09  117.949997  119.430000  117.940002  118.989998   33387600   
9097  2017-01-10  118.769997  119.379997  118.300003  119.110001   24462100   
9098  2017-01-11  118.739998  119.930000  118.599998  119.750000   27588600   
9099  2017-01-12  118.900002  119.300003  118.209999  119.250000   27086200   
9100  2017-01-13  119.110001  119.620003  118.809998  119.040001   26111900   
9101  2017-01-17  118.339996  120.239998  118.220001  120.000000   34439800   
9102  2017-01-18  120.000000  120.500000  119.709999  119.989998   23713000   
9103  2017-01-19  119.400002  120.089996  119.370003  119.779999   25597300   
9104  2017-01-20  120.449997  120.449997  119.730003  120.000000   32597900   
9105  2017-01-23  120.000000  120.809998  119.769997  120.080002   22050200   
9106  2017-01-24  119.550003  120.099998  119.500000  119.970001   23211000   
9107  2017-01-25  120.419998  122.099998  120.279999  121.879997   32377600   
9108  2017-01-26  121.669998  122.440002  121.599998  121.9400

In [97]:
calc_sma(aapl, timespan, ["Close", "Volume"], merge_result=True)

Date        Open        High         Low       Close     Volume  \
0     1980-12-12   28.750000   28.875000   28.750000   28.750000  117258400   
1     1980-12-15   27.375001   27.375001   27.250000   27.250000   43971200   
2     1980-12-16   25.375000   25.375000   25.250000   25.250000   26432000   
3     1980-12-17   25.875000   25.999999   25.875000   25.875000   21610400   
4     1980-12-18   26.625000   26.750000   26.625000   26.625000   18362400   
5     1980-12-19   28.249999   28.374999   28.249999   28.249999   12157600   
6     1980-12-22   29.625000   29.750000   29.625000   29.625000    9340800   
7     1980-12-23   30.874999   30.999999   30.874999   30.874999   11737600   
8     1980-12-24   32.500000   32.624999   32.500000   32.500000   12000800   
9     1980-12-26   35.500001   35.625001   35.500001   35.500001   13893600   
10    1980-12-29   36.000000   36.124999   36.000000   36.000000   23290400   
11    1980-12-30   35.249999   35.249999   35.125000   35.125000   17220000   
12    1980-12-31   34.250000   34.250000   34.125000   34.125000    8937600   
13    1981-01-02   34.499999   34.750001   34.499999   34.499999    5415200   
14    1981-01-05   33.875001   33.875001   33.750001   33.750001    8932000   
15    1981-01-06   32.375000   32.375000   32.250000   32.250000   11289600   
16    1981-01-07   30.999999   30.999999   30.874999   30.874999   13921600   
17    1981-01-08   30.375001   30.375001   30.250001   30.250001    9956800   
18    1981-01-09   31.874999   32.000001   31.874999   31.874999    5376000   
19    1981-01-12   31.874999   31.874999   31.625000   31.625000    5924800   
20    1981-01-13   30.625000   30.625000   30.500000   30.500000    5762400   
21    1981-01-14   30.625000   30.750000   30.625000   30.625000    3572800   
22    1981-01-15   31.250001   31.500000   31.250001   31.250001    3516800   
23    1981-01-16   31.125001   31.125001   30.999999   30.999999    3348800   
24    1981-01-19   32.875001   33.000001   32.875001   32.875001   10393600   
25    1981-01-20   32.000001   32.000001   31.874999   31.874999    7520800   
26    1981-01-21   32.500000   32.749999   32.500000   32.500000    3976000   
27    1981-01-22   32.875001   33.125000   32.875001   32.875001    8887200   
28    1981-01-23   32.875001   33.000001   32.749999   32.749999    2805600   
29    1981-01-26   32.375000   32.375000   32.250000   32.250000    6160000   
...          ...         ...         ...         ...         ...        ...   
9090  2016-12-29  116.449997  117.110001  116.400002  116.730003   15039500   
9091  2016-12-30  116.650002  117.199997  115.430000  115.820000   30586300   
9092  2017-01-03  115.800003  116.330002  114.760002  116.150002   28781900   
9093  2017-01-04  115.849998  116.510002  115.750000  116.019997   21118100   
9094  2017-01-05  115.919998  116.860001  115.809998  116.610001   22193600   
9095  2017-01-06  116.779999  118.160004  116.470001  117.910004   31751900   
9096  2017-01-09  117.949997  119.430000  117.940002  118.989998   33387600   
9097  2017-01-10  118.769997  119.379997  118.300003  119.110001   24462100   
9098  2017-01-11  118.739998  119.930000  118.599998  119.750000   27588600   
9099  2017-01-12  118.900002  119.300003  118.209999  119.250000   27086200   
9100  2017-01-13  119.110001  119.620003  118.809998  119.040001   26111900   
9101  2017-01-17  118.339996  120.239998  118.220001  120.000000   34439800   
9102  2017-01-18  120.000000  120.500000  119.709999  119.989998   23713000   
9103  2017-01-19  119.400002  120.089996  119.370003  119.779999   25597300   
9104  2017-01-20  120.449997  120.449997  119.730003  120.000000   32597900   
9105  2017-01-23  120.000000  120.809998  119.769997  120.080002   22050200   
9106  2017-01-24  119.550003  120.099998  119.500000  119.970001   23211000   
9107  2017-01-25  120.419998  122.099998  120.279999  121.879997   32377600   
9108  2017-01-26  121.669998  122.440002  121.599998  121.9400

In [98]:
calc_bollinger(aapl, timespan, ["Close", "Volume"], merge_result=True, scaler=2)

Date        Open        High         Low       Close     Volume  \
0     1980-12-12   28.750000   28.875000   28.750000   28.750000  117258400   
1     1980-12-15   27.375001   27.375001   27.250000   27.250000   43971200   
2     1980-12-16   25.375000   25.375000   25.250000   25.250000   26432000   
3     1980-12-17   25.875000   25.999999   25.875000   25.875000   21610400   
4     1980-12-18   26.625000   26.750000   26.625000   26.625000   18362400   
5     1980-12-19   28.249999   28.374999   28.249999   28.249999   12157600   
6     1980-12-22   29.625000   29.750000   29.625000   29.625000    9340800   
7     1980-12-23   30.874999   30.999999   30.874999   30.874999   11737600   
8     1980-12-24   32.500000   32.624999   32.500000   32.500000   12000800   
9     1980-12-26   35.500001   35.625001   35.500001   35.500001   13893600   
10    1980-12-29   36.000000   36.124999   36.000000   36.000000   23290400   
11    1980-12-30   35.249999   35.249999   35.125000   35.125000   17220000   
12    1980-12-31   34.250000   34.250000   34.125000   34.125000    8937600   
13    1981-01-02   34.499999   34.750001   34.499999   34.499999    5415200   
14    1981-01-05   33.875001   33.875001   33.750001   33.750001    8932000   
15    1981-01-06   32.375000   32.375000   32.250000   32.250000   11289600   
16    1981-01-07   30.999999   30.999999   30.874999   30.874999   13921600   
17    1981-01-08   30.375001   30.375001   30.250001   30.250001    9956800   
18    1981-01-09   31.874999   32.000001   31.874999   31.874999    5376000   
19    1981-01-12   31.874999   31.874999   31.625000   31.625000    5924800   
20    1981-01-13   30.625000   30.625000   30.500000   30.500000    5762400   
21    1981-01-14   30.625000   30.750000   30.625000   30.625000    3572800   
22    1981-01-15   31.250001   31.500000   31.250001   31.250001    3516800   
23    1981-01-16   31.125001   31.125001   30.999999   30.999999    3348800   
24    1981-01-19   32.875001   33.000001   32.875001   32.875001   10393600   
25    1981-01-20   32.000001   32.000001   31.874999   31.874999    7520800   
26    1981-01-21   32.500000   32.749999   32.500000   32.500000    3976000   
27    1981-01-22   32.875001   33.125000   32.875001   32.875001    8887200   
28    1981-01-23   32.875001   33.000001   32.749999   32.749999    2805600   
29    1981-01-26   32.375000   32.375000   32.250000   32.250000    6160000   
...          ...         ...         ...         ...         ...        ...   
9090  2016-12-29  116.449997  117.110001  116.400002  116.730003   15039500   
9091  2016-12-30  116.650002  117.199997  115.430000  115.820000   30586300   
9092  2017-01-03  115.800003  116.330002  114.760002  116.150002   28781900   
9093  2017-01-04  115.849998  116.510002  115.750000  116.019997   21118100   
9094  2017-01-05  115.919998  116.860001  115.809998  116.610001   22193600   
9095  2017-01-06  116.779999  118.160004  116.470001  117.910004   31751900   
9096  2017-01-09  117.949997  119.430000  117.940002  118.989998   33387600   
9097  2017-01-10  118.769997  119.379997  118.300003  119.110001   24462100   
9098  2017-01-11  118.739998  119.930000  118.599998  119.750000   27588600   
9099  2017-01-12  118.900002  119.300003  118.209999  119.250000   27086200   
9100  2017-01-13  119.110001  119.620003  118.809998  119.040001   26111900   
9101  2017-01-17  118.339996  120.239998  118.220001  120.000000   34439800   
9102  2017-01-18  120.000000  120.500000  119.709999  119.989998   23713000   
9103  2017-01-19  119.400002  120.089996  119.370003  119.779999   25597300   
9104  2017-01-20  120.449997  120.449997  119.730003  120.000000   32597900   
9105  2017-01-23  120.000000  120.809998  119.769997  120.080002   22050200   
9106  2017-01-24  119.550003  120.099998  119.500000  119.970001   23211000   
9107  2017-01-25  120.419998  122.099998  120.279999  121.879997   32377600   
9108  2017-01-26  121.669998  122.440002  121.599998  121.9400

In [100]:
sp500 = load_raw_frame("^GSPC")

In [117]:
sp500_i = sp500.set_index("Date")
aapl_i = aapl.set_index("Date")

In [118]:
joined_df = pd.merge(aapl_i.loc[:,["Close"]], sp500_i.loc[:,["Close"]], how="left", left_index=True, right_index=True)

joined_df.head()

Close_x     Close_y
Date                           
1980-12-12   28.750  129.229996
1980-12-15   27.250  129.449997
1980-12-16   25.250  130.600006
1980-12-17   25.875  132.889999
1980-12-18   26.625  133.000000

In [178]:
timewindow_beta(aapl, sp500, ["Close"], 10)

timewindow_beta_10_Close
Date                                
1980-12-12                       NaN
1980-12-15                       NaN
1980-12-16                       NaN
1980-12-17                       NaN
1980-12-18                       NaN
1980-12-19                       NaN
1980-12-22                       NaN
1980-12-23                       NaN
1980-12-24                       NaN
1980-12-26                  0.602909
1980-12-29                  0.462633
1980-12-30                  0.362929
1980-12-31                  0.272490
1981-01-02                  0.251707
1981-01-05                  0.193142
1981-01-06                 -0.035484
1981-01-07                 -0.001231
1981-01-08                  0.225218
1981-01-09                  0.319368
1981-01-12                  0.368983
1981-01-13                  0.629617
1981-01-14                  0.873622
1981-01-15                  1.018467
1981-01-16                  1.410476
1981-01-19                  0.828619
1981-01-20                 -0.015581
1981-01-21                 -0.262831
1981-01-22                 -0.851694
1981-01-23                 -1.096102
1981-01-26                 -1.198070
...                              ...
2016-12-29                  4.232536
2016-12-30                 13.649676
2017-01-03                 15.201580
2017-01-04                  9.538927
2017-01-05                  8.271823
2017-01-06                 10.394651
2017-01-09                  6.450195
2017-01-10                  5.587053
2017-01-11                  5.475822
2017-01-12                  4.915533
2017-01-13                  4.514116
2017-01-17                  1.458870
2017-01-18                  0.079905
2017-01-19                 -0.646198
2017-01-20                 -2.802713
2017-01-23                 -2.087190
2017-01-24                 -1.096606
2017-01-25                       NaN
2017-01-26                       NaN
2017-01-27                       NaN
2017-01-30                       NaN
2017-01-31                       NaN
2017-02-01                       NaN
2017-02-02                       NaN
2017-02-03                       NaN
2017-02-06                       NaN
2017-02-07                       NaN
2017-02-08                       NaN
2017-02-09                       NaN
2017-02-10                       NaN

[9120 rows x 1 columns]

In [126]:
1980-12-26     0.602909
1980-12-29     0.462633
1980-12-30     0.362929
1980-12-31     0.272490
1981-01-02     0.251707
1981-01-05     0.193142



SyntaxError: invalid syntax (<ipython-input-126-9cb3997a7f82>, line 1)

In [195]:
_i = 10

s = joined_df.iloc[ 0:10, :]["Close_x"]
m = joined_df.iloc[ 0:10, :]["Close_y"]

slope, intercept, r_value, p_value, std_err = stats.linregress(s.values,m.values)
slope

0.60290920052919772

In [181]:
timewindow_alpha(aapl, sp500, ["Close"], 10)

timewindow_beta_10_Close
Date                                
1980-12-12                       NaN
1980-12-15                       NaN
1980-12-16                       NaN
1980-12-17                       NaN
1980-12-18                       NaN
1980-12-19                       NaN
1980-12-22                       NaN
1980-12-23                       NaN
1980-12-24                       NaN
1980-12-26                115.725489
1980-12-29                120.045095
1980-12-30                123.315977
1980-12-31                126.354180
1981-01-02                127.135722
1981-01-05                129.387471
1981-01-06                137.394066
1981-01-07                136.179306
1981-01-08                128.372000
1981-01-09                124.999134
1981-01-12                123.178736
1981-01-13                114.740313
1981-01-14                107.020322
1981-01-15                102.518674
1981-01-16                 90.409048
1981-01-19                108.391856
1981-01-20                134.178299
1981-01-21                141.581753
1981-01-22                160.037691
1981-01-23                167.556352
1981-01-26                170.504536
...                              ...
2016-12-29               1767.624234
2016-12-30                667.265814
2017-01-03                486.016184
2017-01-04               1147.796843
2017-01-05               1295.590680
2017-01-06               1048.530974
2017-01-09               1507.534491
2017-01-10               1607.479948
2017-01-11               1619.789574
2017-01-12               1686.387224
2017-01-13               1735.102745
2017-01-17               2097.505419
2017-01-18               2261.988822
2017-01-19               2347.690245
2017-01-20               2605.589404
2017-01-23               2519.443764
2017-01-24               2402.195440
2017-01-25                       NaN
2017-01-26                       NaN
2017-01-27                       NaN
2017-01-30                       NaN
2017-01-31                       NaN
2017-02-01                       NaN
2017-02-02                       NaN
2017-02-03                       NaN
2017-02-06                       NaN
2017-02-07                       NaN
2017-02-08                       NaN
2017-02-09                       NaN
2017-02-10                       NaN

[9120 rows x 1 columns]